pkg/controller/init_frontend.go

```diff
@@ -1,21 +1,53 @@
 package controller
 
 import (
-	devtypes "github.com/longhorn/go-iscsi-helper/types"
+	"fmt"
 
+	devtypes "github.com/longhorn/go-iscsi-helper/types"
 	"github.com/longhorn/longhorn-engine/pkg/frontend/rest"
 	"github.com/longhorn/longhorn-engine/pkg/frontend/socket"
 	"github.com/longhorn/longhorn-engine/pkg/frontend/tgt"
 	"github.com/longhorn/longhorn-engine/pkg/types"
 )
 
-var (
-	Frontends = map[string]types.Frontend{}
+const (
+	defaultScsiTimeout       = 60 // SCSI device timeout, unit: seconds
+	defaultIscsiAbortTimeout = 15 // unit: seconds
+
+	defaultEngineReplicaTimeout = 8  // unit: seconds
+	minEngineReplicaTimeout     = 8  // unit: seconds
+	maxEngineReplicaTimeout     = 30 // unit: seconds
+
 )
 
-func init() {
-	Frontends["rest"] = rest.New()
-	Frontends["socket"] = socket.New()
-	Frontends[devtypes.FrontendTGTBlockDev] = tgt.New(devtypes.FrontendTGTBlockDev)
-	Frontends[devtypes.FrontendTGTISCSI] = tgt.New(devtypes.FrontendTGTISCSI)
+func NewFrontend(frontendType string, iscsiTargetRequestTimeout int64) (types.Frontend, error) {
+	switch frontendType {
+	case "rest":
+		return rest.New(), nil
+	case "socket":
+		return socket.New(), nil
+	case devtypes.FrontendTGTBlockDev:
+		return tgt.New(devtypes.FrontendTGTBlockDev, defaultScsiTimeout, defaultIscsiAbortTimeout, iscsiTargetRequestTimeout), nil
+	case devtypes.FrontendTGTISCSI:
+		return tgt.New(devtypes.FrontendTGTISCSI, defaultScsiTimeout, defaultIscsiAbortTimeout, iscsiTargetRequestTimeout), nil
+	default:
+		return nil, fmt.Errorf("unsupported frontend type: %v", frontendType)
+	}
+}
+
+func DetermineEngineReplicaTimeout(timeout int64) int64 {
+	if timeout < minEngineReplicaTimeout ||
+		timeout > maxEngineReplicaTimeout {
+		return defaultEngineReplicaTimeout
```

Comment:
Suggest a having a log about using the default timeout, because the value provided is not valid.

Reply:
Addressed. Thank you.